### ISSUE TITLE
Optimal hyperparameters and optimal flow design

### ISSUE THREAD

**Issue by dodreh** [VincentStimper/normalizing-flows#69]
Hello! Thank you for excellent framework!

For my bachelor thesis I have to adapt SAILOR (Link to project: https://github.com/malicd/sailor).  I should implement NF instead of Gaussian Mixture Model. 
Short about SAILOR: SAILOR is unsupervised algorithm for domain adoption. It uses GMM to compress big dataset and to train the model on target dataset. 

How I used NF in SAILOR:
Architecture of transformation/flow: 
Number of flows = 4
![Screenshot from 2024-12-26 16-06-13](https://github.com/user-attachments/assets/72896fa5-940d-4cd5-ad6d-0ac3f0278a07)
I am not sure if the number of hidden layers is optimally chosen or not?

number of epochs = 10 000
learning rate = 0.000002

Problem: I have tried to design flows in different ways (flows = 32, 16, 8, 4;  changing lr and epochs many times, diff. number of hidden layers, different number of units in hidden layers and so on), but no version gave me good results. During the fitting the model in data I have most of the time case that my loss value is "inf" (or "Nan"). A value for loss starts from 365 and decreasing, and in one moment it starts to print "inf" or "nan". I have also printed loss in diagram (number of epochs = 15 000) and at around 2500 and 3000 epoch the diagram shows two very big spikes, after it continues to decrease. 
Question: Is there some standard procedure how to choose hyperparameters  (lr, number of epochs) and how to design flows? If no, can you give me advice how to create optimal flows with optimal parameters (lr, epochs)? Is 15 000 epochs too big or I can consider it as normal value for epochs?

Thank you in advance!

**Comment by dodreh** [comment]
P.S. Here are two screenshots of loss diagram:
![Screenshot from 2024-12-26 16-31-54](https://github.com/user-attachments/assets/3a1aa842-bc44-418f-b087-2163f6a3314f)
![Screenshot from 2024-12-26 16-30-32](https://github.com/user-attachments/assets/06808642-b27c-4a80-8632-e343379274a1)

**Comment by karanchahal-nv** [comment]
I have the same issue seeing these spikes in loss and then stabilizing down. Did you get a chance to find the root case and fix it ?

**Comment by dodreh** [comment]
Hi!

If I remember correctly I have experimented with hyperparameters and found
some combination that gives Okay results. If your loss function converges
to some value and your model gives good results on target dataset (test
dataset) you can ignore that spikes.

Kind regards
Milan
